src/common/snippets/src/pass/split_dimension_m.cpp

```diff
@@ -31,45 +31,55 @@ bool SplitDimensionM::is_supported_matmul(const std::shared_ptr<const ov::Node>&
     const size_t lower_bound = optimal_parallelism_work_amount / batch_dim;
-    if (lower_bound * batch_dim == optimal_parallelism_work_amount && m_dim % lower_bound == 0) {
-        splited.first = lower_bound;
-        splited.second = m_dim / lower_bound;
-        OPENVINO_ASSERT(splited.first * splited.second == m_dim, "Incorrect dimension M splitting!");
-        return splited;
-    }
+    if (lower_bound * batch_dim == optimal_parallelism_work_amount && m_dim % lower_bound == 0)
+        return std::make_pair(lower_bound, m_dim / lower_bound);
 
     // Ideal case #2: M is divisible by optimal parallel work amount, and the new_m_dim is big enough
     // In this case, each thread will execute the Snippets kernel 'batch_dim' times
     if (m_dim % optimal_parallelism_work_amount == 0) {
         const auto new_m_dim = m_dim / optimal_parallelism_work_amount;
         const size_t min_kernel_m = 64;
-        if (new_m_dim >= min_kernel_m) {
-            splited.first = optimal_parallelism_work_amount;
-            splited.second = new_m_dim;
-            OPENVINO_ASSERT(splited.first * splited.second == m_dim, "Incorrect dimension M splitting!");
-            return splited;
-        }
+        if (new_m_dim >= min_kernel_m)
+            return std::make_pair(optimal_parallelism_work_amount, new_m_dim);
     }
 
+    return std::make_pair(1, m_dim);
+}
+
+std::pair<size_t, size_t> SplitDimensionM::compute_conservative_heuristic(size_t batch_dim, size_t m_dim, size_t optimal_parallelism_work_amount) {
+    std::pair<size_t, size_t> splited = { 1, m_dim };
     const size_t upper_bound = utils::div_up(2 * optimal_parallelism_work_amount, batch_dim);
     for (size_t divisor_0 = upper_bound - 1; divisor_0 > 1; divisor_0--) {
         size_t divisor_1 = m_dim / divisor_0;
-        if (divisor_1 * divisor_0 == m_dim) {
-            splited.first = divisor_0;
-            splited.second = divisor_1;
-            break;
-        }
+        if (divisor_1 * divisor_0 == m_dim)
+            return divisor_0 * batch_dim >= optimal_parallelism_work_amount ? std::make_pair(divisor_0, divisor_1) : splited;
     }
-    OPENVINO_ASSERT(splited.first * splited.second == m_dim, "Incorrect dimension M splitting!");
     return splited;
 }
 
+std::pair<size_t, size_t> SplitDimensionM::compute_aggressive_heuristic(size_t batch_dim, size_t m_dim, size_t optimal_parallelism_work_amount) {
+    constexpr size_t min_kernel_m = 32;
+    std::pair<size_t, size_t> best_result = {1, m_dim};
+    for (size_t divisor = 2; divisor < std::sqrt(m_dim); ++divisor) {
+        if (m_dim % divisor != 0)
+            continue;
+        if (divisor >= min_kernel_m)
+            return std::make_pair(m_dim / divisor, divisor);
+        const size_t m_kernel = m_dim / divisor;
+        if (m_kernel >= min_kernel_m) {
+            best_result.first = divisor;
+            best_result.second = m_kernel;
+        }
```

Comment:
The main thing I want to point out is that this heuristic maximizes `m_batch` (= minimizes `m_kernel`) + has a limitation `m_kernel >= min_kernel_m`. In other words, we try to find `m_kernel` bigger than `min_kernel_m` and at the same time as close as possible to this value.
So `batch_dim * m_dim/divisor >= optimal_parallelism_work_amount` is not sufficient criteria. Moreover, we can start this heuristic when `batch_dim` is already bigger than `optimal_parallelism_work_amount`. For the motivation of this strategy, please refer to the function's description:

> Splits m_dim to minimize kernel_m in order to reduce waiting time for idle threads at the last parallel loop iteration.

My logic is structured in the following way (taking into account that divisor is ascending) for the splitting candidates:

- if divisor is more than `min_kernel_m`, (a) strategy is used, and we can guarantee that this is the most optimal implementation from `m_kernel` minimization perspective.
- if divisor is less than `min_kernel_m`, (b) strategy is used. But it is not guaranteed that the current `m_kernel = m_dim / divisor` is minimized: one of the next divisor from `(divisor, sqrt(m_dim))` interval can be more optimal.

Alternatively, I can implement the same logic via 2 `for`s with different traversal directions:

1. `for (size_t divisor = min_kernel_m; divisor < std::sqrt(m_dim); ++divisor)`
2. `for (size_t divisor = min_kernel_m - 1; divisor > 1; --divisor)`

> Sometimes these `min_kernel_m` and `optimal_parallelism_work_amount` can be mutually exclusive, so we should think carefully which is more important. I guess that the parallel work amount should be prioritized, so the approach from 3 should be used.

This is true. But the current heuristic covers the most important cases (big shapes in SD topology), at least on the machines where these changes were tested. And we agreed offline that we need to limit these changes' impact on other topologies. 
Anyway, if I have time on validation, I will try to further tune heuristic to cover the described situation

Reply:
Ok, I see, thanks for the detailed explanation 🙂
It's probably a good idea to add a corresponding comment, it might save us quite some time in the future.